src/main/com/yetanalytics/datasim/sim.clj

```diff
@@ -1,323 +1,337 @@
-            alignment)))
-
-(s/def ::skeleton
-  (s/map-of ::xapi/agent-id
-            :skeleton/statement-seq))
+  (let [actor-alignment-ids (set [actor-id group-id role])]
+    (reduce
+     (fn [alignment-map {component-iri :component :as alignment}]
+       (update alignment-map
+               component-iri
+               (fnil update-alignment {:weight 0.0 :count 0})
+               alignment))
+     {}
+     (for [{alignment-maps :alignments
+            alignment-id   :id} alignments
+           :when (actor-alignment-ids alignment-id)
+           alignment alignment-maps]
+       alignment))))
+
+;; Timestamp helpers
+
+(defn- timezone->region ^ZoneRegion [tz]
+  (t/zone-id tz))
+
+(defn- timestamp->millis [ts]
+  (.toEpochMilli (t/instant ts)))
+
+(defn- drop-statements-from-time
+  "Drop any `statements` whose `:timestamp-ms` metadata comes after
+   `from-ms`."
+  [from-ms statements]
+  (drop-while
+   (fn [statement]
+     (>= from-ms (-> statement meta :timestamp-ms)))
+   statements))
+
+;; Time/probability sequence helpers
+
+;; Right now we are using common ARMA settings, this may change
+(def common-arma
+  {:phi   [0.5 0.2]
+   :theta []
+   :std   0.25
+   :c     0.0})
+
+(defn- arma-seq [seed]
+  (ts/arma-seq (assoc common-arma :seed seed)))
+
+(defn- lunch-hour-seq
+  "Map `minute-of-day-seq` into a sequence of `1.0`, if the corresponding
+   minute of the day is between 12:00 and 13:00 (the \"lunch hour\"),
+   and `-1.0`, if it is any other time."
+  [minute-of-day-seq]
+  (map (fn [min-of-day]
+         (if (<= 720 min-of-day 780) 1.0 -1.0))
+       minute-of-day-seq))
+
+(defn- arma-time-seqs->prob-mask-seq
+  "Derive a lazy sequence of probability mask values for actor events.
+   
+   The mask seq can be plotted as a time series. One will see a
+   generally-sinusoidal value, as `day-night-seq` is sinusoidal,
+   with the highest y-value during the night and lowest during the day;
+   random variation is introduced by `group-arma-seq` and the y-value is
+   fixed at `1.0` during the \"lunch hour\" (12:00 to 13:00)
+   thanks to `(lunch-hour-seq min-of-day-seq)`.
+   
+   These values will be inverted since each mask value will be subtracted
+   from each value in `actor-arma-seq` in order to form each actor
+   event probability."
+  [arma-seq day-night-seq min-of-day-seq]
+  (map max
+       arma-seq
+       day-night-seq
+       (lunch-hour-seq min-of-day-seq)))
+
+(defn- clamp-probability [x]
+  (maths/min-max 0.0 x 1.0))
+
+(defn- arma-mask-seqs->prob-seq
+  "Subtract each value of `arma-seq` by its respective `prob-mask-seq`
+   value, then use that value to derive a probability value in `[0,1]`.
+   Note that a higher `prob-mask-seq` value will result in a lower probability."
+  [arma-seq prob-mask-seq]
+  (map (fn [arma-val prob-mask-val]
+         (-> (- arma-val prob-mask-val) ; higher mask val -> lower prob
+             (/ 2) ; decrease general range from [-1, 1] to [-0.5, 0.5]
```

Comment:
Does this have a marked effect on the number of statements generated in a time-limited sim?

Reply:
First of all this was just refactored behavior; the division by 2 was always there. I'd imagine without it you'd get more probabilities that are either exactly 0 or exactly 1; how exactly this will affect statement gen I don't know, but I assume it would have some sort of effect.